src/ast/passes/codegen_llvm.cpp

```diff
@@ -2582,6 +2587,50 @@ void CodegenLLVM::generateProbe(Probe &probe,
         func_type, section_name, current_attach_point_->address, index);
 }
 
+void CodegenLLVM::visit(Subprog &subprog)
+{
+  std::vector<llvm::Type *> arg_types;
+  std::transform(subprog.args->begin(),
+                 subprog.args->end(),
+                 std::back_inserter(arg_types),
+                 [this](SubprogArg *arg) { return b_.GetType(arg->type); });
+  FunctionType *func_type = FunctionType::get(b_.GetType(subprog.return_type),
+                                              arg_types,
+                                              0);
+
+  Function *func = Function::Create(
+      func_type, Function::InternalLinkage, subprog.name(), module_.get());
+  BasicBlock *entry = BasicBlock::Create(module_->getContext(), "entry", func);
+  b_.SetInsertPoint(entry);
+
+  variables_.clear();
+  ctx_ = nullptr;
```

Comment:
Would it make sense to hoist this out to a helper function? I think `generateProbe()` shares the `variables_.clear()` bit. It would help self-document the intent and make future extensions less error prone

Reply:
Parts of it are shared, but parts of it are also different, subprogs process arguments, probes do not, probe do additional setup subprogs don't.